### PR TITLE
[mlir][ArmSME] Lower transfer_write + transpose to vertical store

### PR DESCRIPTION
**Pull request by c-rhodes**
This patch extends the lowering of vector.transfer_write in
VectorToArmSME to support in-flight transpose via SME vertical store.

### CODE REVIEW

File: mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp [hunk 1]
```diff
@@ -136,13 +136,31 @@ struct TransferReadToArmSMELowering
...
+///
+/// Example 1: op with identity permutation map to horizontal
+///            arm_sme.tile_store:
+///
+///   vector.transfer_write %vector, %source[%c0, %c0]
+///     {in_bounds = [true, true]} : vector<[16]x[16]xi8>, memref<?x?xi8>
 ///
 /// is converted to:
 ///
 ///   arm_sme.tile_store %vector, %source[%c0, %c0] : memref<?x?xi8>,
```
Comment (banach-space) [inline mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:149]: Given that this is now always passing `layout` to the builder, shouldn't this be printed as:
```
arm_sme.tile_store %vector, %source[%c0, %c0] layout<horizontal> : memref<?x?xi8>, vector<[16]x[16]xi8>
```

Comment (c-rhodes) [reply mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:149]: horizontal layout is the default and isn't printed, but I could add it to be explicit if you think that's clearer?

File: mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp [hunk 2]
```diff
@@ -153,12 +171,35 @@ struct TransferWriteToArmSMELowering
...
+    AffineMap map = writeOp.getPermutationMap();
+    if (map.isIdentity())
+      layout = arm_sme::TileSliceLayout::Horizontal;
+    else if (map == AffineMap::get(map.getNumDims(), 0, {d1, d0},
+                                   writeOp.getContext()))
+      layout = arm_sme::TileSliceLayout::Vertical;
+    else
+      return rewriter.notifyMatchFailure(writeOp,
+                                         "unsupported permutation map");
+
```
Comment (banach-space) [inline mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:199]: Not a fan of nested "else if" :)
```suggestion
     bool isTranspose = (map == AffineMap::get(map.getNumDims(), 0, {d1, d0},
                                               transferReadOp.getContext()));
    if (!map.isIdentity() && !isTranspose)
        return rewriter.notifyMatchFailure(transferReadOp,
                                           "unsupported permutation map");
     arm_sme::TileSliceLayout layout =
         isTranspose ? arm_sme::TileSliceLayout::Vertical
                     : arm_sme::TileSliceLayout::Horizontal;
```

Apologies for the formatting - GitHub is not great for these things.

**Review by banach-space — COMMENTED** [review]
Makes sense, just some usual nitpicking from me :)

Thanks!

File: mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp (hunk 1 above)
Comment (MacDue) [reply mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp:149]: My preference is the omit expected default values (and I think horizontal would be expected) :)

File: mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-transfer-write-2d.mlir [hunk 3]
```diff
@@ -116,6 +135,25 @@ func.func @entry() {
...
+  call @transfer_write_2d_transposed(%A, %c0, %c0) : (memref<?x?xf32>, index, index) -> ()
+  call @load_and_print(%A, %c0, %c0) : (memref<?x?xf32>, index, index) -> ()
+
+  // 5. Reload 4. + store + transpose but with mask (nrows=4, ncols=2).
+  // The mask applies after permutation
+  // CHECK-LABEL: TILE BEGIN:
+  // CHECK-NEXT: ( 0, 0, 20, 30
+  // CHECK-NEXT: ( 0, 0, 21, 31
+  // CHECK-NEXT: ( 20, 21, 0, 0
+  // CHECK-NEXT: ( 30, 31, 0, 0
```
[the remark this answers is not in the document]
Comment (c-rhodes) [reply mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-transfer-write-2d.mlir]: correct, took me a minute to figure it out again as well :) added a comment to clarify.

File: mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp (hunk 2 above)
Comment (c-rhodes) [reply mlir/lib/Conversion/VectorToArmSME/VectorToArmSME.cpp]: that is nicer 👍

**Review by banach-space — APPROVED** [review]
LGTM, thanks for addressing my comments!

**Pull request merged**